### DISCUSSION TITLE
Time management service refuses processing interaction

### DISCUSSION THREAD

**Discussion by Cheng-Yuan-0124** [eclipse/mosaic#243] [Q&A]
Hi,

Currently, I am trying to integrate the MOSAIC with CARLA simulator. The integration of ambassador and interaction have been implemented but there is an issue I am encountering. When the function `receiveInteraction` in the `AbstractFederateAmbassador` script successfully adds my interaction into the `interactionQueue`, the time management service refuses or ignores my interaction sometimes. This kind of issue happening **_randomly_** in a run. Sometimes happening after running several hours, sometimes happening within few minutes or sometimes never happening in a run. This issue cause the simulation freezing because my script can not receive the repose of the request interaction. Is there any clue or any part I could check to deal with it? 

Thank you very much

**Comment by kschrab** [comment]
Hi, this is rather hard to debug for us. I could only thing of interactions having outdated timestamps. Did you make sure that MOSAIC RTI is controlling the simulation time in Carla and not vice-versa? Have you tried to step trough the RTI for debugging, e.g., through `AbstractFederateAmbassador#receiveInteraction`?`

Nevertheless, we also working on a coupling with CARLA and hope to release a initial version of it at the end of the year.

**Reply by Cheng-Yuan-0124** [reply]
Hi, thanks for the replying. I've found the issue causing simulation freezing at the end which is exact coming from `SequentialTimeManagement`. The issue is when two of the federate events in the `EfficientPriorityQueue` with same `lookahead`,  `priority` and `requestedTime`, they will be sorted by the `federateId` (or other variables). The picture below shows the situation I mention, the federateId `carla` will be the next event to be polled from the queue but it will be always added back to the queue at the first index which is prior than the `sumo` federateId. As a result, the `sumo` federateId will never be processed then causing freezing. Do you have any idea about how to properly deal with this situation or is there anything I did in the wrong way? Thanks!
![Image from iOS](https://user-images.githubusercontent.com/32282492/182161573-dc7d2b95-49ec-4069-9877-57207371e8cf.jpg)
.

**Reply by kschrab** [reply]
Which of both events should be executed at first? Have you already tried passing a higher priority than `2` when calling `rti.requestAdvanceTime(time, lookahead, priority)` from your CarlaAmbassador/Federate?

**Reply by Cheng-Yuan-0124** [reply]
In the case I mention above, the `interactionQueue` in CARLA ambassador doesn't have any data so the first event should be executed is sumo federate event. I've tried passing a higher priority from my CarlaAmbassador but it will have kind of the same freezing situation when first event should be CARLA but stuck in sumo federate event because the `interactionQueue` in SUMO ambassador is empty. The current solution I have is adding 1 nano second such as `rti.requestAdvanceTime(time + executeTimes, lookahead, priority)` in CarlaAmbassador which `executeTimes` equals to 0 at beginning but being counted when `requestAdvanceTime` being processed and reset it when CarlaAmbassador received the respond, then the CARLA federate event will not be added back to the first index in the event queue.

I am not sure it is or not a correct way to use MOSAIC, hope you could give me some advise such as how you handling this situation when you are implementing the CARLA coupling work
